### PR TITLE
Adapter and config refactor lvl1

### LINKED ISSUE #443: Revit_Toolkit: Refactor BHoM <-> Revit type queries into data structures
#### Description:
At the moment there is a few methods that define the relationship between BHoM and Revit types and categories:
- [ ] `BH.UI.Revit.Engine.Query.BHoMTypes`
- [ ] `BH.UI.Revit.Engine.Query.BuiltInCategories`
- [ ] `BH.UI.Revit.Engine.Query.BuiltInCategory`
- [ ] `BH.UI.Revit.Engine.Query.FramingType`
- [x] `BH.UI.Revit.Engine.Query.RevitType`
- [ ] `BH.UI.Revit.Engine.Query.RevitTypes`

We need to structure and document these as tables and reflect in code in a form of appropriate data structures.

### LINKED ISSUE #448: Revit_Toolkit: Better handling of ConvertBeforePushError in BH.UI.Revit.Adapter.Create
#### Description:
At the moment there is a hacky check of whether the BHoMObject can be converted:

https://github.com/BHoM/Revit_Toolkit/blob/ca828c02ee5a5bd0eb787634f3ff837e5d27cae0/Adapter_Revit_UI/CRUD/Create.cs#L115-L124

Would be good to handle in a more civilised manner, possibly as a part of #452.

### LINKED ISSUE #452: Revit_Toolkit: Break Read and Create methods into smaller entities
#### Description:
Currently the methods are hardly readable and often do a bit of a ping-pong. It would be good to partition them into smaller sections so that each would correspond to a single item in the execution diagram.

### LINKED ISSUE #514: Revit_Toolkit: Simplify Settings in favour of RevitConfig and PushType/PullType
#### Description:

Current structure of the Adapter allows for using `ActionConfig`, `PullType` and `PushType` - it would be good to simplify the settings sitting in `BH.oM.Adapters.Revit.Settings` and pass as much of them as possible via the adapter objects.

Ideally it would be good to have 3 sources of settings:
- adapter settings (global)
- `RevitPullConfig` and `RevitPushConfig` - splitting of `RevitConfig` required

### LINKED ISSUE #529: Revit_Toolkit: Implement PushType and PullType
#### Description:

Implement and replace the existing source of info on push/pull type.

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Closes #448 
Closes #514 
Closes #529 
Partially closes #452, although this one will need further refinement (to be done right after merging this PR not to mess things up).
Partially closes #443 by squashing `BHoMTypes` method.

### Test files

[#448](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue448%2DConvertBeforePushError)
[#514 - pull with various requests](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2FRevit%5FToolkit%2DIssue514%2DSettingsAndConfig)
[Macro push beams and columns - DeleteThenCreate part added](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPush%2FPhysical)
[Macro pull](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPull%2FStructure)

### Changelog

- `PullSettings` class has been entirely deleted, replaced with `RevitPullConfig` to large extent
- `PushSettings` class has been entirely deleted, replaced with `RevitPushConfig` to large extent
- `UpdateTagsSettings` and `GeneralSettings` have been merged into `RevitSettings`, the latter currently have 6 properties:
    - `ConnectionSettings` - untouched recently
    - `FamilyLoadSettings` - untouched recently
    - `MapSettings` - untouched recently
    - `TagsParameterName` - to be merged into `MapSettings` in long term (#533)
    - `DistanceTolerance` and `AngleTolerance`
- All convert methods have been refactored so that:
    - They do not carry `PullSettings` and `PushSettings` arguments any more
    - They take `RevitSettings` argument that contains MapSettings, Tolerance etc.
    - They take a `refObjects` dictionary argument that contains converted objects coupled with ids of source objects to avoid converting same objects many times - before the dictionary was a property of `PushSettings` and `PullSettings`, now it is carried separately to enable using it outside adapter actions
- `CopyCustomData` setting has been removed - custom data is always copied on convert
- `Read` method has been further simplified to accommodate the above changes and to make the code less convoluted
- `Push` adapter action and `Create` method have been preliminarily adjusted to accommodate the above changes as well as to embrace `PushType` - currently only `CreateOnly` and `DeleteThenCreate` are allowed - prototype of `UpdateOnly` to follow in the next PRs

### Additional comments
`RevitFilePreview` is temporarily not supported as I wanted to move it out of the core `Create` method - the code will be reused in the next PRs. I am letting myself disable this functionality as it is quite niche while it was contaminating the core methods.

Immediate follow-up to this PR will be:
- reenabling `RevitFilePreview` - ideally to be handled in a similar manner as any other `IBHoMObject`
- prototype of `UpdateOnly` #546
- introducing Remove adapter action #535

Precedence of this PR over #530 has been agreed with @tiagogrossi

### CODE REVIEW

**Comment by pawelbaran** [comment]
@FraserGreenroyd could you please beat me with your compliance stick in case something was wrong?
@MajaLindroth @kThorsager could you please check if your scripts still work?
@IsakNaslundBh @alelom  would be fantastic if you could have a look into the changes made to the core adapter part (but I know both of you are superbusy)

Help with merging this PR will be highly appreciated - this is just one (biggest!) step towards the core-level refactoring that I would like to finish before 3.1 release - one or two more is still needed...

File: Revit_Engine/Create/Settings/RevitSettings.cs [hunk 1]
```diff
@@ -38,7 +38,7 @@ public static partial class Create
         [Input("familyLoadSettings", "FamilyLoad Settings for Revit Adapter")]
         [Input("generalSettings", "General Settings for Revit Adapter")]
```
Comment (FraserGreenroyd) [inline Revit_Engine/Create/Settings/RevitSettings.cs]: This input is not mapped to any attribute, it needs removing

**Comment by pawelbaran** [comment]
@MajaLindroth could you please provide some more specific info? Which parts did go missing? Is it one type of object or random? Could you maybe share the minimum test file that highlights the issues? Thanks!

**Comment by pawelbaran** [comment]
All comments and change requests have been processed.

@MajaLindroth I have migrated your models one level down in directory, it is now [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FRevit%5FToolkit%2F%5FMacro%2FPull%2FEnvironment%2FFrom%20MajaLindroth) - please feel free to move around or rename if you like.

**Pull request merged**